crypto/bn/rsaz_exp_x2.c

```diff
@@ -0,0 +1,484 @@
+void RSAZ_exp52x20_x2(BN_ULONG *res, const BN_ULONG *base,
+                      const BN_ULONG *exp[2], const BN_ULONG *m,
+                      const BN_ULONG *rr, const BN_ULONG k0[2]);
+/*
+ * AMM = Almost Montgomery Multiplication.
+ * _256 suffix means it uses AVX512VL 256-bit width IFMA.
+ */
+void RSAZ_amm52x20_x2_256(BN_ULONG *out, const BN_ULONG *a,
+                          const BN_ULONG *b, const BN_ULONG *m,
+                          const BN_ULONG k0[2]);
+/* AMS = Almost Montgomery Squaring. */
+void RSAZ_ams52x20_x2_256(BN_ULONG *out, const BN_ULONG *a,
+                          const BN_ULONG *m, const BN_ULONG k0[2]);
+void extract_multiplier_2x20_win5(BN_ULONG *red_Y,
+                                  const BN_ULONG *red_table,
+                                  int red_table_idx, int tbl_idx);
+
+/*
+ * Parallel Montgomery modular exponentiation of 2-primes with moduli of the
+ * same bit size, optimized with AVX512 ISA.
+ *
+ * Input and output parameters for each exponentiation are independent and
+ * denoted here by index |i|, i = 1..2.
+ *
+ * Input and output are all in regular 2^64 radix.
+ *
+ * Each moduli shall be |factor_size| bit size.
+ *
+ * NOTE: currently only 2x1024 case is supported.
+ *
+ *  [out] res|i|      - result of modular exponentiation: array of qword values
+ *                      in regular (2^64) radix. Size of array shall be enough
+ *                      to hold |factor_size| bits.
+ *  [in]  base|i|     - base
+ *  [in]  exp|i|      - exponent
+ *  [in]  m|i|        - moduli
+ *  [in]  rr|i|       - Montgomery parameter R^2
+ *  [in]  k0_|i|      - Montgomery parameter k0 = -1/m|i| mod 2^64
+ *  [in]  factor_size - moduli bit size
+ *
+ * \return 0 in case of failure,
+ *         1 in case of success.
+ */
+int RSAZ_mod_exp_avx512_x2(BN_ULONG *res1,
+                           const BN_ULONG *base1,
+                           const BN_ULONG *exp1,
+                           const BN_ULONG *m1,
+                           const BN_ULONG *rr1,
+                           BN_ULONG k0_1,
+                           BN_ULONG *res2,
+                           const BN_ULONG *base2,
+                           const BN_ULONG *exp2,
+                           const BN_ULONG *m2,
+                           const BN_ULONG *rr2,
+                           BN_ULONG k0_2,
+                           int factor_size)
+{
+    int ret = 0;
+
+    /*
+     * Number of word-size (BN_ULONG) digits to store exponent in redundant
+     * representation.
+     */
+    int exp_digits = number_of_digits(factor_size + 2, 52);
+    BN_ULONG *base1_red, *m1_red, *rr1_red, *res1_red;
+    BN_ULONG *base2_red, *m2_red, *rr2_red, *res2_red;
+    BN_ULONG *storage = NULL;
+    BN_ULONG *storage_aligned = NULL;
+    BN_ULONG storage_len_bytes = 8 * exp_digits * sizeof(BN_ULONG);
+
+    /* AMM = Almost Montgomery Multiplication */
+    AMM52 AMM = NULL;
+    EXP52_x2 EXPx2 = NULL;
```

Comment:
caps variable names is not OpenSSL style

Reply:
Ok, changed.